netket/vqs/exact/expect.py

```diff
@@ -63,20 +64,22 @@ def expect(vstate: ExactState, Ô: DiscreteOperator) -> Stats:  # noqa: F811
 
 
 @dispatch
-def expect_and_grad(
+def expect_and_forces(
     vstate: ExactState,
     Ô: DiscreteOperator,
-    use_covariance: TrueT,
     *,
     mutable: Any,
 ) -> Tuple[Stats, PyTree]:
+    if isinstance(Ô, Squared):
+        raise NotImplementedError("expect_and_forces not yet implemented for `Squared`")
+
     _check_hilbert(vstate, Ô)
 
     O = sparsify(Ô)
```

Comment:
what does this do? can we add comments?

Reply:
`sparsify` calls `O.to_sparse()` in order to evaluate the expectations via full summation. The function caches the last five evaluations in order to reduce the number of calls to `to_sparse`. I think the comment and docstring on `sparsify` are the best place to mention that, at least I'd keeping that information there once rather than repeating it at the places where `sparsify` is called:
https://github.com/netket/netket/blob/cdb85c77922efb7625f4281658c96f7326fb36aa/netket/vqs/exact/expect.py#L39-L46